src/charm.py

```diff
@@ -44,21 +44,36 @@ def __init__(self, *args):
         self.framework.observe(self.on['grafana-source'].relation_changed,
                                self._on_grafana_changed)
 
-        if self._stored.provider_ready:
+        if self.provider_ready:
             self.prometheus_provider = MonitoringProvider(self,
                                                           'monitoring', 'prometheus', self.version)
             self.framework.observe(self.prometheus_provider.on.targets_changed,
-                                   self._on_config_changed)
+                                   self._on_scrape_targets_changed)
```

Comment:
I would prefer to avoid partial construction even if it means more sophisticated logic later.

Reply:
Not sure what you mean. The whole point of `self.provider_ready` is to avoid "partial construction". There is a [ticket](https://github.com/canonical/prometheus-operator/issues/41) to explore alternative options.

Reply:
Also not that `ProviderBase` has a `ready()` method that was perhaps intended to deal with partial construction. I am yet to explore how this can be integrated into this library.

Reply:
I meant that the charm class is constructed conditionally: at different times it may have different composition.

Reply:
Yes. That I will try and address in a future enhancement as it may not be trivial. I have to explore it.